File: C++/countBinarySubstrings.cpp
/**
 * @File    : countBinarySubstrings.cpp
 * @Brief   : 计数二进制子串
 * @Link    : https://leetcode-cn.com/problems/count-binary-substrings/
 * @Author  : Wei Li
 * @Date    : 2021-07-28
*/

#include <iostream>
#include <string>

class Solution
{
public:
    int countBinarySubstrings(std::string s)
    {
        int pre = 0, cur = 1, count = 0;
        for (int i = 1; i < s.length(); ++i)
        {
            if (s[i] == s[i - 1])
            {
                ++cur;
            }
            else
            {
                pre = cur;
                cur = 1;
            }
            if (pre >= cur)
            {
                ++count;
            }
        }
        return count;
    }
};

// ------------------------------
int main(int argc, char **argv)
{
    // std::string str = "00110011";
    std::string str = "10101";

    Solution solution;
    auto num_binary_substring = solution.countBinarySubstrings(str);
    std::cout << "The solution of this problem is " << num_binary_substring << std::endl;

    return 0;
}

File: Python/countBinarySubstrings.py
#!/usr/bin/env python3
# encoding: utf-8

"""
@Filename: countBinarySubstrings.py
@Function: 计数二进制子串
@Link: https://leetcode-cn.com/problems/count-binary-substrings/
@Python Version: 3.8
@Author: Wei Li
@Date：2021-07-28
"""


class Solution:
    def countBinarySubstrings(self, s: str) -> int:
        count = 0
        for i in range(len(s) - 1):
            if int(s[i]) ^ int(s[i + 1]) == 1:
                index = 0
                while i - index >= 0 and i + 1 + index <= len(s) - 1:
                    if s[i - index] == s[i] and s[i + 1] == s[i + 1 + index]:
                        count += 1
                        index += 1
                    else:
                        break
        return count


# ----------------------------------
if __name__ == "__main__":
    str = "00110011"

    solution = Solution()
    num_binary_substring = solution.countBinarySubstrings(str)
    print(f"The solution of this problem is : {num_binary_substring}")
